app/forms/firewall-rules-edit.tsx

```diff
@@ -5,36 +5,53 @@
  *
  * Copyright Oxide Computer Company
  */
+import { useNavigate, type LoaderFunctionArgs } from 'react-router-dom'
+
 import {
+  apiQueryClient,
   firewallRuleGetToPut,
   useApiMutation,
   useApiQueryClient,
-  type VpcFirewallRule,
+  usePrefetchedApiQuery,
 } from '@oxide/api'
 
 import { SideModalForm } from '~/components/form/SideModalForm'
-import { useForm, useVpcSelector } from '~/hooks'
+import { getVpcSelector, useFirewallRuleSelector, useForm, useVpcSelector } from '~/hooks'
+import { invariant } from '~/util/invariant'
+import { pb } from '~/util/path-builder'
 
 import {
   CommonFields,
   valuesToRuleUpdate,
   type FirewallRuleValues,
 } from './firewall-rules-create'
 
-type EditFirewallRuleFormProps = {
-  onDismiss: () => void
-  existingRules: VpcFirewallRule[]
-  originalRule: VpcFirewallRule
+EditFirewallRuleForm.loader = async ({ params }: LoaderFunctionArgs) => {
+  const vpcSelector = getVpcSelector(params)
+
+  await apiQueryClient.prefetchQuery('vpcFirewallRulesView', {
+    query: vpcSelector,
+  })
+
+  return null
 }
 
-export function EditFirewallRuleForm({
-  onDismiss,
-  existingRules,
-  originalRule,
-}: EditFirewallRuleFormProps) {
+export function EditFirewallRuleForm() {
+  const { vpc, project, rule } = useFirewallRuleSelector()
   const vpcSelector = useVpcSelector()
   const queryClient = useApiQueryClient()
 
+  const { data } = usePrefetchedApiQuery('vpcFirewallRulesView', {
+    query: { project, vpc },
+  })
+
+  const originalRule = data.rules.find((r) => r.name === rule)
+
+  invariant(originalRule, 'Firewall rule must exist')
```

Comment:
This isn't quite right because this is generic server error and we want to 404 here. I will make the loader 404 if the rule isn't there.